pdns/dnsdistdist/dnsdist-opentelemetry.cc

```diff
@@ -0,0 +1,279 @@
+      otTrace.resource_spans.at(0).scope_spans.at(0).spans.end(),
+      d_postActivationSpans->begin(),
+      d_postActivationSpans->end());
+  }
+  return otTrace;
+#endif
+}
+
+std::string Tracer::getOTProtobuf() const
+{
+#ifdef DISABLE_PROTOBUF
+  return 0;
+#else
+  // TODO: Should we close all spans?
+  return getTracesData().encode();
+#endif
+}
+
+SpanID Tracer::addSpan([[maybe_unused]] const std::string& name)
+{
+#ifdef DISABLE_PROTOBUF
+  return 0;
+#else
+  return addSpan(name, SpanID{});
+#endif
+}
+
+SpanID Tracer::addSpan([[maybe_unused]] const std::string& name, [[maybe_unused]] const SpanID& parentSpanID)
+{
+#ifdef DISABLE_PROTOBUF
+  return 0;
+#else
+  auto spanID = pdns::trace::randomSpanID();
+  if (d_activated) {
+    d_postActivationSpans->push_back({
+      .trace_id = d_traceid,
+      .span_id = spanID,
+      .parent_span_id = parentSpanID,
+      .name = name,
+      .start_time_unix_nano = pdns::trace::timestamp(),
+    });
+    return spanID;
+  }
+
+  // We're not activated, so we are in pre-activation.
+  if (d_preActivationSpans == nullptr) {
+    // This is the first time addEvent is called for this Tracer instance
+    d_preActivationSpans = std::make_unique<std::vector<preActivationSpanInfo>>();
+  }
+
+  d_preActivationSpans->push_back({
+    .name = name,
+    .span_id = spanID,
+    .parent_span_id = parentSpanID,
+    .start_time_unix_nano = pdns::trace::timestamp(),
+    .end_time_unix_nano = 0,
+  });
+
+  d_lastSpanID = spanID;
+
+  return spanID;
+#endif
+}
+
+// TODO: Figure out what to do with duplicate keys
+void Tracer::setTraceAttribute([[maybe_unused]] const std::string& key, [[maybe_unused]] const AnyValue& value)
+{
+#ifndef DISABLE_PROTOBUF
+  if (!d_activated) {
+    // TODO: throw an error? Log an error? We don't want to collect any data before we are are actually tracing
+    return;
+  }
+  if (d_attributes == nullptr) {
+    d_attributes = std::make_unique<std::vector<pdns::trace::KeyValue>>();
+  }
+  d_attributes->push_back({key, value});
+#endif
+}
+
+void Tracer::closeSpan([[maybe_unused]] const SpanID& spanID)
+{
+#ifndef DISABLE_PROTOBUF
+  if (d_activated) {
+    auto spanIt = std::find_if(
+      d_postActivationSpans->rbegin(),
+      d_postActivationSpans->rend(),
+      [spanID](const pdns::trace::Span& span) { return span.span_id == spanID; });
+    if (spanIt != d_postActivationSpans->rend()) {
+      if (spanIt->end_time_unix_nano == 0) {
+        spanIt->end_time_unix_nano = pdns::trace::timestamp();
+      }
+      return;
+    }
+  }
+
+  auto spanIt = std::find_if(
+    d_preActivationSpans->rbegin(),
+    d_preActivationSpans->rend(),
+    [spanID](const preActivationSpanInfo& span) { return span.span_id == spanID; });
+  if (spanIt != d_preActivationSpans->rend() && spanIt->end_time_unix_nano == 0) {
+    spanIt->end_time_unix_nano = pdns::trace::timestamp();
+    return;
+  }
+#endif
+}
+
+// TODO: Figure out what to do with duplicate keys
```

Comment:
What's the impact of having duplicate keys? If we want to prevent them we can use a `unordered_map` instead of a `vector`, but I don't know if it is worth the trouble.

Reply:
As the actual data in the protobuf is also a vector of key-values, I'll keep it this way. It would be good to see how jaeger handles this.